src/librustc/infer/error_reporting/nice_region_error/static_impl_trait.rs

```diff
@@ -52,17 +53,24 @@ impl<'a, 'tcx> NiceRegionError<'a, 'tcx> {
                         }) => name.to_string(),
                         _ => "'_".to_owned(),
                     };
-                    if let Ok(snippet) = self.tcx().sess.source_map().span_to_snippet(return_sp) {
-                        err.span_suggestion(
-                            return_sp,
-                            &format!(
-                                "you can add a constraint to the return type to make it last \
+                    let fn_return_span = return_ty.unwrap().1;
+                    if let Ok(snippet) =
+                        self.tcx().sess.source_map().span_to_snippet(fn_return_span) {
+                        // only apply this suggestion onto functions with
+                        // explicit `impl Trait` return.
+                        // e.g. `fn foo() -> impl Trait {}`
+                        if !fn_return_span.is_empty() {
```

Comment:
I think instead of `is_empty`, we should invoke [`is_desugaring`](https://doc.rust-lang.org/nightly/nightly-rustc/syntax_pos/span_encoding/struct.Span.html#method.is_desugaring) or perhaps [`desugaring_kind`](https://doc.rust-lang.org/nightly/nightly-rustc/syntax_pos/span_encoding/struct.Span.html#method.desugaring_kind) -- something like `fn_return_span.is_desugaring(DesugaringKind::Async)`?

Reply:
arguably we could be conservative and just avoid for *any* desugaring